sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,280 @@
+from ..base import BaseEstimator, TransformerMixin
+from ..utils.validation import check_is_fitted
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+
+def nca_vectorized_oracle(X, y, n_components, loss, threshold=0.0):
+    dx = X[:, np.newaxis, :] - X[np.newaxis, :, :]  # n_samples x n_samples x n_features
+    outer = dx[:, :, np.newaxis, :] * dx[:, :, :, np.newaxis]  # n_samples x n_samples x n_features x n_features
```

Comment:
Yes, L is always either square of rectangular with `n_components < n_features`. The opposite doesn't make sense, since L's rank is bounded by `n_features`.

What's the benefit of propagating L inside? In case of vectorized optimizer we would lose the ability to cache outer products. It might be useful in semivectorized optimizer, but the effect seems useless: we'll be computing `L (x - y) (x - y)^T` products which doesn't seem to be of any help in case of square L (though, I can compute it efficiently since I already have `L(x-y)` from probabilities).